File: app/src/main/java/com/example/sarada/moviereviews/ReviewAdapter.java
package com.example.sarada.moviereviews;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.example.sarada.moviereviews.models.Review;

import java.util.List;

public class ReviewAdapter extends RecyclerView.Adapter<ReviewAdapter.MyReviewViewHolder> {

    private Context mContext;
    private List<Review> reviewList;

    public ReviewAdapter(Context mContext, List<Review> reviewList){
        this.mContext = mContext;
        this.reviewList = reviewList;

    }

    @Override
    public ReviewAdapter.MyReviewViewHolder onCreateViewHolder(ViewGroup viewGroup, int i){
        View view = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.review_card, viewGroup, false);
        return new ReviewAdapter.MyReviewViewHolder(view);

    }

    @Override
    public void onBindViewHolder(final ReviewAdapter.MyReviewViewHolder viewHolder, int i){
        viewHolder.author.setText(reviewList.get(i).getAuthor());
        viewHolder.content.setText(reviewList.get(i).getContent());

    }

    @Override
    public int getItemCount(){

        return reviewList.size();

    }

    public class MyReviewViewHolder extends RecyclerView.ViewHolder{
        public TextView author, content;

        public MyReviewViewHolder(View view){
            super(view);
            author = (TextView) view.findViewById(R.id.author_name);
            content = (TextView) view.findViewById(R.id.content);
        }
    }
}

File: app/src/main/java/com/example/sarada/moviereviews/presentation/adapters/ReviewAdapter.kt
package com.example.sarada.moviereviews.presentation.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestOptions
import com.example.sarada.moviereviews.R
import com.example.sarada.moviereviews.presentation.adapters.ReviewAdapter.MyReviewViewHolder
import com.example.sarada.moviereviews.databinding.ReviewCardBinding
import com.example.sarada.moviereviews.models.datac.review.Review

class ReviewAdapter(
    private val mContext: Context
) : ListAdapter<Review, MyReviewViewHolder>(ReviewDiffCallback()) {

    override fun onCreateViewHolder(parent: ViewGroup, i: Int): MyReviewViewHolder {
        return MyReviewViewHolder.from(mContext, parent)
    }

    override fun onBindViewHolder(holder: MyReviewViewHolder, position: Int) {
        val item = getItem(position)
        holder.bind(item)
    }

    class MyReviewViewHolder private constructor(
        private val mContext: Context,
        val binding: ReviewCardBinding
    ) : RecyclerView.ViewHolder(binding.root) {

        fun bind(
            item: Review
        ) {
            binding.review = item
            binding.executePendingBindings()

            var avatarPath: String? = null
            if (item.authorDetails.avatarPath != null) {
                avatarPath =
                    if (item.authorDetails.avatarPath!!.length < 40)
                        "https://image.tmdb.org/t/p/w500${item.authorDetails.avatarPath}"
                    else item.authorDetails.avatarPath
            }
            Glide.with(mContext)
                .load(avatarPath)
                .apply(RequestOptions().placeholder(R.drawable.ic_outline_account_circle_24))
                .apply(RequestOptions().circleCrop())
                .into(binding.authorAvatar)

            var isTextViewExpanded = false
            binding.content.setOnClickListener {
                if (!isTextViewExpanded) {
                    binding.content.maxLines = 100
                    isTextViewExpanded = true
                } else {
                    binding.content.maxLines = 2
                    isTextViewExpanded = false
                }
            }
        }

        companion object {
            fun from(mContext: Context, parent: ViewGroup): MyReviewViewHolder {
                val layoutInflater = LayoutInflater.from(parent.context)
                val binding = ReviewCardBinding.inflate(layoutInflater, parent, false)
                return MyReviewViewHolder(mContext, binding)
            }
        }
    }
}

class ReviewDiffCallback : DiffUtil.ItemCallback<Review>() {
    override fun areItemsTheSame(oldItem: Review, newItem: Review): Boolean {
        return oldItem.author == newItem.author
    }

    override fun areContentsTheSame(oldItem: Review, newItem: Review): Boolean {
        return oldItem == newItem
    }

}